### PR TITLE
[Task-5] Back-end introduced

### PR DESCRIPTION
**Pull request by PetkevichVladislav**
Prepare

Install and configure [eslint-plugin-react-hooks](https://www.npmjs.com/package/eslint-plugin-react-hooks). It will help you identify mistakes when using hooks.
Clone the [backend repository](https://github.com/VarvaraZadnepriak/MoviesAPI.ReactJS).
Navigate to the cloned repository folder and run "npm install" to install dependencies.
Run "npm start" command to start the backend server. It will start on https://localhost:4000.
Open Swagger API docs: http://localhost:4000/api-docs.
 

Implement movie list page

Now it's time to wire our components up and build the real app.

Create a "MovieListPage" component. You will place all the markup and logic for the main page here. Using useState hook define several states for the component:

search query
sort criterion
active genre
movie list (array)
selected movie (default to undefined or null)
others if necessary
Render components that you've previously built to create the movie list page. It should look similar to the [design prototype](https://www.figma.com/file/fKGjrOqR6nJe6LYJopGCZ8/%5BCDP%5D-Home-Task-%E2%80%93-React-v1). Pixel-perfect design is not necessary, but overall composition and placement should look the same.

Pass search query state to the SearchForm component. Modify the state when the search form is submitted.
If selected movie is defined, instead of SearchForm render MovieDetails and pass the selected movie in props.
Pass static genre list and an active genre to the GenreSelect component. Modify the active genre state when other genres are selected.
Pass current sort criterion to SortControl. Modify sort criterion state when the user changes sorting.
Render MovieTile for every movie in state. Pass movie info to the MovieTile. You can define a mock movie array for now, so that your array is not empty and you can test results. Handle click events on MovieTile and update the selected movie state when a movie is selected.
As a result of this step, you should have a movie list page, with all controls working but without backend integration.

 

Implement effects

By using useEffect hook and [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) or [axios](https://axios-http.com/), implement the following effect:

When search query, sort criterion or active genre changes, make a request to the backend to get movies corresponding to the new parameters. On response, update movie list with response data. Don't forget to abort previous request in case the user quickly updates several parameters.

All functionality of the app except forms should work now.

 

Write end-to-end tests

You've implemented a lot of functionality, composed several components to achieve a working piece of software. Now spend some time on covering this behavior with a reliable end-to-end test. Cover search functionality, sorting and switching genres, selecting a movie and returning back to search.

### CODE REVIEW

File: src/components/Menu/Menu.tsx [hunk 1]
```diff
@@ -1,22 +1,19 @@
...
+import { SortControl } from "../SortControl/SortControl";
+import { Genre } from "../../models/enum/MenuGenre";
+import { SortOption } from "../../models/enum/SortOption";
 
 const Menu = () => {
-    const [itemNames, setItemNames] = useState<string[]>(["Action", "Comedy", "Drama", "Sci-Fi", "Thriller", "Western"]);
-    const [preselectedItemName, setSelectedItemName] = useState<string>("Sci-Fi");
-    const onSelect = useCallback((query: string) => { console.log(query); }, []);
-    const onSortOptionChange = useCallback((currentOption:SortOption) => console.log(currentOption), []);
+    const genres = useMemo(() => Object.values(Genre), []);
```
Comment (iZamza) [inline src/components/Menu/Menu.tsx]: types

File: src/components/MovieCard/MovieCard.scss [hunk 2]
```diff
@@ -47,6 +47,19 @@
...
         right: 15px;
     }
+
+    &__image{
+        width: 322px;
+        height: 455px;
+    }
+
+    &__image-container{
+        background-color: #FFFFFF;
```
Comment (iZamza) [inline src/components/MovieCard/MovieCard.scss]: variable

File: src/components/MovieCard/MovieCard.tsx [hunk 3]
```diff
@@ -1,36 +1,34 @@
...
+    onCardClick: () => void;
     movieMenu: MenuModel;
 }
 
-export const MovieCard: FC<IMovieCard> = (movie: IMovieCard) => {
+export const MovieCard: FC<IMovieCard> = ({card, id, onCardClick, movieMenu}) => {
     const [isMouseHoveredOnCard, setIsMouseHoveredOnCard] = useState<boolean>(false);
     const [isPopupMenuOpened, setIsPopupMenuOpened] = useState<boolean>(false);
-    const [menuItems, setMenuItems] = useState([
+    const menuItems = useMemo( () => [
```
Comment (iZamza) [inline src/components/MovieCard/MovieCard.tsx]: space

File: src/components/MovieDetails/MovieDetails.tsx [hunk 4]
```diff
@@ -1,46 +1,39 @@
...
 
 export interface IMovieDetails {
-    imageUrl: string;
-    movieName: string;
-    movieGenres: string[];
-    releaseYear: string;
-    rating: number;
-    duration: string;
-    description: string;
+    movie?: MovieModel | null | undefined;
```
Comment (iZamza) [inline src/components/MovieDetails/MovieDetails.tsx]: `movie?: MovieModel | null`

File: src/components/MovieListPage/MovieListPage.tsx [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+}
+
+export const MovieListPage: FC = () => {
+    const [searchQuery, setSearchQuery] = useState<string>("");
+    const [sortCriterion, setSortCriterion] = useState<SortOption>(SortOption.ReleaseDate);
+    const [selectedGenre, setSelectedGenre] = useState<Genre>(Genre.All);
+    const updateSelectedGenre = useCallback((genre: Genre) => {
+        setSelectedGenre(genre);
+        setSearchQuery("");
+    },[setSearchQuery, setSelectedGenre]);
```
Comment (iZamza) [inline src/components/MovieListPage/MovieListPage.tsx]: enter, and keep all useCallbacks or useMemos after all useStates

**Review by iZamza — APPROVED** [review]
Nice job!

File: src/components/Menu/Menu.tsx (hunk 1 above)
Comment (PetkevichVladislav) [reply src/components/Menu/Menu.tsx]: Type added, thank you for your comment.

File: src/components/MovieCard/MovieCard.tsx (hunk 3 above)
Comment (PetkevichVladislav) [reply src/components/MovieCard/MovieCard.tsx]: Space removed, thank you for your comment.

File: src/components/MovieDetails/MovieDetails.tsx (hunk 4 above)
Comment (PetkevichVladislav) [reply src/components/MovieDetails/MovieDetails.tsx]: Input parameter simplyfied, thank you for your comment.

File: src/components/MovieListPage/MovieListPage.tsx (hunk 5 above)
Comment (PetkevichVladislav) [reply src/components/MovieListPage/MovieListPage.tsx]: Hooks ordered, Thank you for your comment.

File: src/components/MovieCard/MovieCard.scss (hunk 2 above)
Comment (PetkevichVladislav) [reply src/components/MovieCard/MovieCard.scss]: Vriable for empty image color introduced, thank you for your comment.

**Pull request merged**